kedro/framework/cli/starters.py

```diff
@@ -592,11 +586,12 @@ def _fetch_config_from_file(config_path: str) -> dict[str, str]:
             f"Failed to generate project: could not load config at {config_path}."
         ) from exc
 
-    return config
+    # The return type defined is more specific than the "Any" type config return from yaml.safe_load
+    return config  # type: ignore[no-any-return]
 
 
 def _fetch_config_from_user_prompts(
-    prompts: dict[str, Any], cookiecutter_context: OrderedDict
+    prompts: dict[str, Any], cookiecutter_context: OrderedDict | None
```

Comment:
I found this a bit odd and run the same `mypy` check locally, I didn't find mypy complains the original signature.

Here we don't really expect `None` to be passed, I see that you have to add another guard clause to check `if not cookiedcutter_context`. If we apply this logic everywhere, we will almost always check for every single argument for `None`, thus I am curious what make this function a special case.

> kedro/framework/cli/starters.py:393: error: Call to untyped function "_validate_selected_tools" in typed context  [no-untyped-call]
> kedro/framework/cli/starters.py:397: error: Call to untyped function "_validate_regex" in typed context  [no-untyped-call]
> kedro/framework/cli/starters.py:401: error: Call to untyped function "_validate_regex" in typed context  [no-untyped-call]
> kedro/framework/cli/starters.py:404: error: Returning Any from function declared to return "dict[str, Any]"  [no-any-return]
> kedro/framework/cli/starters.py:445: error: Call to untyped function "_safe_load_entry_point" in typed context  [no-untyped-call]
> kedro/framework/cli/starters.py:523: error: Unused "type: ignore" comment  [unused-ignore]
> kedro/framework/cli/starters.py:595: error: Returning Any from function declared to return "dict[str, str]"  [no-any-return]
> kedro/framework/cli/starters.py:635: error: Function is missing a return type annotation  [no-untyped-def]
> kedro/framework/cli/starters.py:643: error: Function is missing a return type annotation  [no-untyped-def]
> kedro/framework/cli/starters.py:705: error: Function is missing a return type annotation  [no-untyped-def]
> kedro/framework/cli/starters.py:742: error: Function is missing a return type annotation  [no-untyped-def]

Reply:
Yeah this one took me a while to fix. The original error is:
```
kedro/framework/cli/starters.py:287: error: Argument "cookiecutter_context" to "_get_extra_context" has incompatible type "OrderedDict[Any, Any] | None"; expected "OrderedDict[Any, Any]"  [arg-type]
```

The problem is that `mypy` can't figure out that `_fetch_config_from_user_prompts` will never be called when `cookiecutter_context` is `None`. I don't know if there's a more elegant solution to fix this.

Reply:
The only fix I could see is changing the condition that calls _fetch_config_from_user_prompts to be `if cookiecutter_context`, which would work given we have a cookiecutter context if and only if we don't have a config path. However, this isn't representative of the actual logic and just exploits the iff condition - I don't think satisfying mypy is worth it. I'm happy to keep `Ordered Dict | None` in the type hint and add a line to the docstring below explaining that the cookiecutter context cannot be None (and hence a guardrail condition has been omitted).

Alternatively we could add a guardrail condition for consistency.

Reply:
> I don't think satisfying mypy is worth it

So would you prefer to add an `type: ignore` here over the current solution?

Reply:
I should clarify, I don't think satisfying `mypy` is worth restructuring how we perform our condition checks in `_get_extra_context `. I'd prefer to leave the type hint as is with a note in the docstring that `cookiecutter_context` is never expected to be None / `_fetch_config_from_user_prompts` will only be called with a `cookiecutter_context`